oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/incrementalstore/IncrementalStoreBuilder.java

```diff
@@ -0,0 +1,152 @@
+
+
+    public enum IncrementalSortStrategyType {
+        /**
+         * Incremental store having nodes updated between initial and final checkpoint
+         */
+
+        INCREMENTAL_FFS_STORE
+    }
+
+    public IncrementalStoreBuilder(File workDir, IndexHelper indexHelper,
+                                   @NotNull String initialCheckpoint, @NotNull String finalCheckpoint) {
+        this.workDir = workDir;
+        this.indexHelper = indexHelper;
+        this.initialCheckpoint = Objects.requireNonNull(initialCheckpoint);
+        this.finalCheckpoint = Objects.requireNonNull(finalCheckpoint);
+    }
+
+    public IncrementalStoreBuilder withPreferredPathElements(Set<String> preferredPathElements) {
+        this.preferredPathElements = preferredPathElements;
+        return this;
+    }
+
+    public IncrementalStoreBuilder withSortStrategyType(IncrementalStoreBuilder.IncrementalSortStrategyType sortStrategyType) {
+        this.sortStrategyType = sortStrategyType;
+        return this;
+    }
+
+    public IncrementalStoreBuilder withPathPredicate(Predicate<String> pathPredicate) {
+        this.pathPredicate = pathPredicate;
+        return this;
+    }
+
+    public IncrementalStoreBuilder withBlobStore(BlobStore blobStore) {
+        this.blobStore = blobStore;
+        return this;
+    }
+
+
+    public IncrementalStore build() throws IOException, CompositeException {
+        logFlags();
+        File dir = createStoreDir();
+
+        switch (sortStrategyType) {
+            case INCREMENTAL_FFS_STORE:
+                NodeStateEntryWriter entryWriter = new NodeStateEntryWriter(blobStore);
+                IncrementalIndexStoreSortStrategy strategy = new IncrementalFlatFileStoreStrategy(
+                        indexHelper.getNodeStore(),
+                        initialCheckpoint,
+                        finalCheckpoint,
+                        dir, preferredPathElements, algorithm, pathPredicate, entryWriter);
+                File metadataFile = strategy.createMetadataFile();
+                File incrementalStoreFile = strategy.createSortedStoreFile();
+                long entryCount = strategy.getEntryCount();
+                IncrementalStore store = new IncrementalFlatFileStore(blobStore, incrementalStoreFile, metadataFile,
+                        new NodeStateEntryReader(blobStore),
+                        unmodifiableSet(preferredPathElements), algorithm);
+                if (entryCount > 0) {
+                    store.setEntryCount(entryCount);
+                }
+                return store;
+        }
+        throw new IllegalStateException("Not a valid sort strategy value " + sortStrategyType);
+    }
+
+    private File createStoreDir() throws IOException {
+        return Files.createTempDirectory(workDir.toPath(), getDirNamePrefix()).toFile();
+    }
+
+    private String getDirNamePrefix() {
+        return INCREMENTAL_STORE_DIR_NAME_PREFIX + sortStrategyTypeString;
```

Comment:
sortStrategyTypeString - this can be null if system property is not set , we should probably use the string rep of sortStrategyType instead.